cpu/sam0_common/periph/flashpage.c

```diff
@@ -45,6 +42,39 @@ void flashpage_write(int page, void *data)
         PAC1->WPCLR.reg = NVMCTRL_PAC_BIT;
     }
 #endif
+}
+
+void flashpage_write_raw(void *target_addr, void *data, size_t len)
+{
+    /* The actual minimal block size for writing is 16B, thus we
+     * assert we write on multiples and no less of that length.
+     */
+    assert(!(len % FLASHPAGE_RAW_BLOCKSIZE));
+
+    /* ensure 4 byte aligned writes */
+    assert(!(((unsigned)target_addr % FLASHPAGE_RAW_ALIGNMENT) ||
+            ((unsigned)data % FLASHPAGE_RAW_ALIGNMENT)));
+
+    uint32_t *page_addr = (uint32_t *)target_addr;
+    uint32_t *data_addr = (uint32_t *)data;
+
+    /* write 4 bytes in one go */
+    len /= 4;
+
+    NVMCTRL->CTRLA.reg = (NVMCTRL_CTRLA_CMDEX_KEY | NVMCTRL_CTRLA_CMD_PBC);
```

Comment:
just one more thing I'd like to clarify: the `write_raw()` function is writing the `NVMCTRL` peripheral, but the `NVMCTRL` peripheral is not unlocked in this function. Shouldn't the `write_raw()` also call `_unlock()`? Seems to me we can not always cound on `flashpage_write()` having been called before and call to `write_raw()`?!

Reply:
True. And actually I was thinking, if we unlock it we should lock it again isn't it?